# https://python.lerner.co.il

Password: boipw!!

# Agenda

1. Loops
2. List
3. Dictionary
4. Files
5. Modules

In [1]:
s = 'abcd'

print(s[0])
print(s[1])
print(s[2])
print(s[3])

a
b
c
d


# Loops

1. `for` loops
2. `while` loops

In [2]:
s = 'abcd'

for one_character in s:
    print(one_character)

a
b
c
d


In [3]:
s = 'abcd'

for one_character in s[2:]:
    print(one_character)

c
d


In [4]:
for counter in 3:
    print('Hooray')

TypeError: 'int' object is not iterable

In [5]:
for counter in range(3):    # use range(n) for n iterations
    print('Hooray')

Hooray
Hooray
Hooray


In [6]:
for counter in range(3):    # range(3) gives me 0, 1, 2
    print(f'{counter} Hooray')

0 Hooray
1 Hooray
2 Hooray


In [7]:
# to end a loop early, use break
# if we've achieved our goal early

for counter in range(10):
    if counter == 5:
        break
    print(counter)

0
1
2
3
4


In [10]:
# to end the current iteration early, use continue
# if there's no reason to continue with the current one

for counter in range(10):
    if counter == 5:
        continue    # go onto the next iteration
    print(counter)

0
1
2
3
4
6
7
8
9


In [11]:
s = 'abcdefghijklmnopqrstuvwxyz'

to_find = 'j'

for one_character in s:
    if one_character == to_find:
        print(f'Found {to_find}! Exiting')
        break
    print(f'Looking for {to_find}.... found {one_character}')

Looking for j.... found a
Looking for j.... found b
Looking for j.... found c
Looking for j.... found d
Looking for j.... found e
Looking for j.... found f
Looking for j.... found g
Looking for j.... found h
Looking for j.... found i
Found j! Exiting


In [12]:
s = 'abcdefghijklmnopqrstuvwxyz'

to_find = 'j'

for one_character in s:
    if one_character != to_find:
        continue
    print(f'Found {to_find}! Exiting')


Found j! Exiting


In [17]:
# s = 'abcdefghijklmnopqrstuvwxyz'
s = 'abcde'

to_find = 'jklmn'

for one_character in s:
    if one_character in to_find: 
        print(f'{one_character} is in {to_find}! Exiting')
        break


In [21]:
# add the digits in s

s = 'a1b2c34'
total = 0

for one_character in s:
    print(f'\tLooking at {one_character}')
    if one_character.isdigit():
        print(f'\t\tFound a digit, adding to total')
        total += int(one_character)
    
print(total)    

	Looking at a
	Looking at 1
		Found a digit, adding to total
	Looking at b
	Looking at 2
		Found a digit, adding to total
	Looking at c
	Looking at 3
		Found a digit, adding to total
	Looking at 4
		Found a digit, adding to total
10


In [22]:
for one_item in range(5):   
    print(one_item)

0
1
2
3
4


DRY -- don't repeat yourself

In [24]:
s = 'abcdefghijklmnopqrstuvwxyz'

for one_letter in s[10:20]:
    print(one_letter)

k
l
m
n
o
p
q
r
s
t


In [25]:
for one_item in range(10, 20):
    print(one_item)

10
11
12
13
14
15
16
17
18
19


In [26]:
for one_item in range(10, 20, 3):
    print(one_item)

10
13
16
19


# Exercise: Digits, vowels, and others

1. Define `digits`, `vowels`, and `others` -- all to be 0
2. Ask the user to enter some text.
3. Go over the text, one character at a time:
    - if it's a digit, add 1 to `digits`
    - if it's a vowel (a, e, i, o, u), add 1 to `vowels`
    - otherwise, add 1 to `others`
4. Print the values of `digits`, `vowels`, and `others`

https://github.com/reuven/boi-autumn-2024|

In [27]:
digits = 0
vowels = 0
others = 0

text = input('Enter text: ').strip()

for one_character in text:
    if one_character in 'aeiou':    # if it's a vowel
        vowels += 1                 #    add 1 to vowels
    elif one_character.isdigit():   # if it's a digit
        digits += 1                 #    add 1 to digits
    else:
        others += 1                 # otherwise, add 1 to others
        
print(f'digits = {digits}')        
print(f'vowels = {vowels}')
print(f'others = {others}')

Enter text: hello !! 123
digits = 3
vowels = 2
others = 7


# `while`

`while` works just like `if`, except that we run the loop body until the test is `False`.

In [28]:
x = 5

while x > 0:
    print(x)
    x -= 1    # x = x - 1

5
4
3
2
1


In [29]:
total = 0

# ask the user to give us numbers until we reach 100

while total < 100:
    s = input('Enter a number: ').strip()

    if not s.isdigit():
        print(f'{s} is not a number; try again')
        continue

    total += int(s)
    print(f'\tCurrent total is {total}')
    
print(total)    
    



Enter a number: 5
	Current total is 5
Enter a number: 20
	Current total is 25
Enter a number: asdvasdfad
asdvasdfad is not a number; try again
Enter a number: 30
	Current total is 55
Enter a number: 1
	Current total is 56
Enter a number: 1
	Current total is 57
Enter a number: 1
	Current total is 58
Enter a number: 1
	Current total is 59
Enter a number: 80
	Current total is 139
139


In [30]:
while True:   
    name = input('Enter your name: ').strip()
    
    if name == '':
        break
        
    print(f'Hello, {name}!')

Enter your name: Reuven
Hello, Reuven!
Enter your name: asdfafadfafd
Hello, asdfafadfafd!
Enter your name: 


# List

We can store *anything* we want in a list. The items we enter stay in their original order. We can modify a list! (Lists are "mutable.")

We define a list with `[]`, with commas (`,`) between elements.

In [31]:
mylist = [10, 20, 30, 40, 50, 60, 70]

# get the length with len
len(mylist)

7

In [32]:
mylist[0]

10

In [33]:
mylist[1]

20

In [34]:
mylist[-1]

70

In [35]:
mylist[2:5]

[30, 40, 50]

In [36]:
for one_item in mylist:
    print(one_item)

10
20
30
40
50
60
70


In [37]:
40 in mylist

True

In [38]:
# lists can change!
# we can change an existing value

mylist[3] = '!!!'
mylist

[10, 20, 30, '!!!', 50, 60, 70]

In [39]:
# we can add new items to the list, using list.append
mylist.append(80)
mylist.append('hello')

mylist

[10, 20, 30, '!!!', 50, 60, 70, 80, 'hello']

In [40]:
# we can remove with list.pop, whose default is to remove from the end
mylist.pop()

'hello'

In [41]:
mylist

[10, 20, 30, '!!!', 50, 60, 70, 80]

In [42]:
mylist.pop()

80

In [43]:
mylist

[10, 20, 30, '!!!', 50, 60, 70]

In [44]:
x = 'abcde'
y = x

x = 'fghij'
y

'abcde'

In [45]:
x = [10, 20, 30]
y = x

x[1] = 9999
x

[10, 9999, 30]

In [46]:
y

[10, 9999, 30]

In [47]:
x = ['abc', 'defgh', 'ij']

for one_item in x:
    print(one_item.upper())

ABC
DEFGH
IJ


In [48]:
type(x)

list

In [49]:
odds = []
evens = []

numbers = [10, 15, 17, 20]

for one_number in numbers:
    if one_number % 2 == 1:
        odds.append(one_number)
    else:
        evens.append(one_number)
        
print(odds)        
print(evens)


[15, 17]
[10, 20]


In [50]:
mylist = ['hello', 'out', 'there']

'hello' in mylist

True

In [51]:
'h' in mylist

False

In [52]:
'h' in mylist[0]

True

In [53]:
for one_item in mylist:
    if 'h' in one_item:
        print(f'Found "h" in {one_item}')

Found "h" in hello
Found "h" in there


# Exercise: Vowels, digits, and others... list edition

1. Define three empty lists (`vowels`, `digits`, `others`).
2. Ask the user to enter text, again and again
    - If we get an empty string (presses ENTER alone), exit
3. Iterate over the text we got from the user, one character at a time
    - If we got a vowel, add it to `vowels`
    - If we got a digit, add it to `digits`
    - Otherwise, add to `others`
    
Example:

    Enter text: hello!
    Enter text: 123
    Enter text: [ENTER]
    vowels: ['e', 'o']
    digits: ['1', '2', '3']
    others: ['h', 'l', 'l', '!']